File: src/algorithms/recursion/nth-fibonacci/solution3.go
package main

// O(n) time | O(1) space
func GetNthFib(n int) int {
	lastTwo := []int{0, 1}
	counter := 3
	for counter <= n {
		nextFib := lastTwo[0] + lastTwo[1]
		lastTwo = []int{lastTwo[1], nextFib}
		counter += 1
	}
	if n > 1 {
		return lastTwo[1]
	}
	return lastTwo[0]
}

File: src/algorithms/recursion/nth-fibonacci/solution3.ts
// O(n) Time | O(1) Space
export function getNthFib(n: number) {
  let lastTwo: [number, number] = [0, 1];
  let counter = 3;
  while (counter <= n) {
    let nextFib = lastTwo[0] + lastTwo[1];
    lastTwo[0] = lastTwo[1];
    lastTwo[1] = nextFib;
    counter++;
  }
  return n > 1 ? lastTwo[1] : lastTwo[0];
}
